ShutdownModeMockServerTest is flaky

```
[INFO] Running com.google.cloud.spanner.pgadapter.ShutdownModeMockServerTest
Error:  Tests run: 3, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.144 s <<< FAILURE! -- in com.google.cloud.spanner.pgadapter.ShutdownModeMockServerTest
Error:  com.google.cloud.spanner.pgadapter.ShutdownModeMockServerTest.testSmartShutdownFollowedByFastShutdown -- Time elapsed: 0.045 s <<< FAILURE!
java.lang.AssertionError: expected:<STOPPING> but was:<TERMINATED>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:120)
	at org.junit.Assert.assertEquals(Assert.java:146)
	at com.google.cloud.spanner.pgadapter.ShutdownModeMockServerTest.testSmartShutdownFollowedByFastShutdown(ShutdownModeMockServerTest.java:202)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junitcore.JUnitCore.run(JUnitCore.java:49)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.createRequestAndRun(JUnitCoreWrapper.java:120)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.executeEager(JUnitCoreWrapper.java:95)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:75)
	at org.apache.maven.surefire.junitcore.JUnitCoreWrapper.execute(JUnitCoreWrapper.java:69)
	at org.apache.maven.surefire.junitcore.JUnitCoreProvider.invoke(JUnitCoreProvider.java:146)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:385)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:507)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:495)
```

--- a/src/main/java/com/google/cloud/spanner/pgadapter/ProxyServer.java
+++ b/src/main/java/com/google/cloud/spanner/pgadapter/ProxyServer.java
@@ -438,7 +438,7 @@ public class ProxyServer extends AbstractApiService {
 
   private void createConnectionHandlersTerminatedLatch() {
     synchronized (this.handlers) {
-      this.allHandlersTerminatedLatch.set(new CountDownLatch(1));
+      this.allHandlersTerminatedLatch.set(new CountDownLatch(handlers.isEmpty() ? 0 : 1));
     }
   }
 

--- a/src/main/java/com/google/cloud/spanner/pgadapter/ProxyServer.java
+++ b/src/main/java/com/google/cloud/spanner/pgadapter/ProxyServer.java
@@ -323,6 +323,7 @@ public class ProxyServer extends AbstractApiService {
   }
 
   void setShutdownMode(ShutdownMode shutdownMode) {
+    logger.log(Level.INFO, "Setting shutdown mode to {0}", shutdownMode);
     this.shutdownMode.set(shutdownMode);
   }
 
@@ -437,7 +438,7 @@ public class ProxyServer extends AbstractApiService {
 
   private void createConnectionHandlersTerminatedLatch() {
     synchronized (this.handlers) {
-      this.allHandlersTerminatedLatch.set(new CountDownLatch(this.handlers.size()));
+      this.allHandlersTerminatedLatch.set(new CountDownLatch(1));
     }
   }
 
@@ -460,7 +461,7 @@ public class ProxyServer extends AbstractApiService {
   void deregister(ConnectionHandler handler) {
     synchronized (this.handlers) {
       this.handlers.remove(handler);
-      if (this.allHandlersTerminatedLatch.get() != null) {
+      if (this.handlers.isEmpty() && this.allHandlersTerminatedLatch.get() != null) {
         this.allHandlersTerminatedLatch.get().countDown();
       }
     }

--- a/src/main/java/com/google/cloud/spanner/pgadapter/ShutdownHandler.java
+++ b/src/main/java/com/google/cloud/spanner/pgadapter/ShutdownHandler.java
@@ -46,6 +46,7 @@ public class ShutdownHandler {
 
   /** Shuts down the proxy server using the given {@link ShutdownMode}. */
   public synchronized void shutdown(@Nonnull ShutdownMode shutdownMode) {
+    Preconditions.checkNotNull(shutdownMode);
     if (this.shutdownMode.get() != null) {
       if (this.shutdownMode.get() == ShutdownMode.SMART
           && (shutdownMode == ShutdownMode.FAST || shutdownMode == ShutdownMode.IMMEDIATE)) {
@@ -55,7 +56,7 @@ public class ShutdownHandler {
       }
       return;
     }
-    this.shutdownMode.set(Preconditions.checkNotNull(shutdownMode));
+    this.shutdownMode.set(shutdownMode);
     this.shutdownThread.start();
   }
 }

--- a/src/test/java/com/google/cloud/spanner/pgadapter/ShutdownModeMockServerTest.java
+++ b/src/test/java/com/google/cloud/spanner/pgadapter/ShutdownModeMockServerTest.java
@@ -58,7 +58,9 @@ public class ShutdownModeMockServerTest extends AbstractMockServerTest {
 
   @After
   public void stopProxyServer() {
-    proxyServer.stopServer();
+    if (proxyServer.state() != State.TERMINATED) {
+      proxyServer.stopServer();
+    }
   }
 
   private String createUrl() {

--- a/src/test/java/com/google/cloud/spanner/pgadapter/ShutdownModeMockServerTest.java
+++ b/src/test/java/com/google/cloud/spanner/pgadapter/ShutdownModeMockServerTest.java
@@ -124,6 +124,48 @@ public class ShutdownModeMockServerTest extends AbstractMockServerTest {
   }
 
   @Test
+  public void testSmartShutdownWithoutConnections() throws SQLException {
+    String sql = "SELECT 1";
+
+    ShutdownHandler shutdownHandler = ShutdownHandler.createForServer(proxyServer);
+    // Verify that the server is running.
+    assertEquals(State.RUNNING, proxyServer.state());
+
+    try (Connection connection = DriverManager.getConnection(createUrl())) {
+      // Verify that the connection is valid.
+      try (ResultSet resultSet = connection.createStatement().executeQuery(sql)) {
+        assertTrue(resultSet.next());
+        assertEquals(1L, resultSet.getLong(1));
+        assertFalse(resultSet.next());
+      }
+    }
+    // Initiate a smart shutdown.
+    shutdownHandler.shutdown(ShutdownMode.SMART);
+    // Wait until the server port is no longer open.
+    Stopwatch stopwatch = Stopwatch.createStarted();
+    while (isPortListening(proxyServer.getLocalPort())
+        && stopwatch.elapsed(TimeUnit.MILLISECONDS) < 2000) {
+      Thread.yield();
+    }
+    // Verify that we cannot open a new connection.
+    SQLException exception =
+        assertThrows(SQLException.class, () -> DriverManager.getConnection(createUrl()));
+    assertEquals(
+        String.format(
+            "Connection to localhost:%d refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.",
+            proxyServer.getLocalPort()),
+        exception.getMessage());
+
+    Stopwatch.createStarted();
+    while (proxyServer.state() == State.STOPPING
+        && stopwatch.elapsed(TimeUnit.MILLISECONDS) < 2000) {
+      Thread.yield();
+    }
+    // Verify that the server terminates.
+    assertEquals(State.TERMINATED, proxyServer.state());
+  }
+
+  @Test
   public void testFastShutdown() throws SQLException {
     String sql = "SELECT 1";
 

